Backend/FcEvaluate.hs

```diff
@@ -0,0 +1,143 @@
+
+-- Error propagation
+smallStep (FcTmApp    e1@(FcTmERROR {}) _) = pure $ Just e1
+smallStep (FcTmTyApp  e1@(FcTmERROR {}) _) = pure $ Just e1
+smallStep (FcTmCaseFc e1@(FcTmERROR {}) _) = pure $ Just e1
+
+-- Useful cases (substitution)
+smallStep (FcTmApp (FcTmAbs x _ body) e2) = Just <$> do
+  newe <- freshenLclBndrs e2
+  pure $ substVar x newe body
+smallStep (FcTmTyApp (FcTmTyAbs a body) ty) = Just <$> do
+  newty <- freshenLclBndrs ty
+  pure $ substVar a newty body
+smallStep (FcTmLet x ty e1 e2) = do
+  newe1 <- freshenLclBndrs $ FcTmLet x ty e1 e1
+  pure <$> Just $ substVar x newe1 e2
+smallStep (FcTmCaseFc scr alts)
+  | FcTmUnit <- scr
+  = matchTheUnit alts
+  | Just (dc, args) <- userDefinedDataConAppMaybe scr
+  = matchTheAlts dc args alts
+
+-- Congruences
+smallStep (FcTmApp e1 e2) = smallStep e1 >>= \case
+  Just e1' -> pure $ Just $ FcTmApp e1' e2
+  Nothing  -> pure Nothing
+smallStep (FcTmTyApp e1 ty) = smallStep e1 >>= \case
+  Just e1' -> pure $ Just $ FcTmTyApp e1' ty
+  Nothing  -> pure Nothing
+smallStep (FcTmCaseFc scr alts) = smallStep scr >>= \case
+  Just scr' -> pure $ Just $ FcTmCaseFc scr' alts
+  Nothing   -> pure $ Nothing
+
+matchTheUnit :: [FcAlt 'Fc] -> EvM (Maybe (FcTerm 'Fc))
+matchTheUnit []                              = error "<unit impossible: pm compiler failed!>"
+matchTheUnit ((FcAltFc FcUnitPat     rhs):_) = pure $ Just rhs
+matchTheUnit ((FcAltFc (FcConPat {}) _  ):_) = error "<unit impossible: elab failed!>"
+
+matchTheAlts :: FcDataCon -> [FcTerm 'Fc] -> [FcAlt 'Fc] -> EvM (Maybe (FcTerm 'Fc))
+matchTheAlts _dc _args []                           = error "<impossible: pm compiler failed!>"
+matchTheAlts _dc _args ((FcAltFc FcUnitPat _rhs):_) = error "<impossible: elab failed!>"
+matchTheAlts  dc  args ((FcAltFc (FcConPat dc' xs) rhs):rest)
+  | dc == dc' = if
+      | length args /= length xs -> error "<impossible: elab failed!>"
+      | otherwise -> do
+          sub <- buildFcTmSubst . zipExact xs <$> mapM freshenLclBndrs args
+          pure $ Just $ substFcTmInTm sub rhs
+  | otherwise = matchTheAlts dc args rest
+
+
+-- collapseProgram :: FcProgram 'Fc -> FcTerm 'Fc
+-- smallStep :: MonadUnique m => FcTerm 'Fc -> m (Maybe (FcTerm 'Fc))
+
+groundTerm :: FcTerm 'Fc -> FcTerm 'Fc
+groundTerm (FcTmTyAbs ty t1) = FcTmTyApp (FcTmTyAbs ty (groundTerm t1)) FcTyUnit
+groundTerm t                 = t
+
+-- | Follows lazy operations semantics
+fullStep :: FcTerm 'Fc -> EvM (Either String (FcTerm 'Fc))
+fullStep t = smallStep t >>= \case
+  Just t'  -> fullStep t'
+  Nothing  -> case t of
+    FcTmERROR err _ -> return $ Left err
+    res             -> return $ Right res
+
+-- | Fully evaluates all branches, by calling fullStep
+fullEval :: FcTerm 'Fc -> EvM (Either String (FcTerm 'Fc))
+fullEval t = fullStep t >>= \case
+  Right (FcTmApp t1 t2) -> fullEval t2 >>= \case
+    Right t2' -> fullStep (FcTmApp t1 t2')
+    Left err  -> return $ Left err
+  result -> return $ result
```

Comment:
You can leave this function as-is for now, but I think that you should explain somewhere about our evaluation strategy (at least a comment).